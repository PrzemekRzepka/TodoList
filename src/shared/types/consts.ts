export const REMOVE_ITEM_TEXT = "REMOVE ITEM";
export const ADD_TEXT = "ADD";
export const TEXT_INPUT_PLACEHOLDER_TEXT = "To do task text";