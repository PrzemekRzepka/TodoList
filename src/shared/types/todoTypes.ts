export type TodoItem = {
    id: string,
    text: string,
    isDone: boolean,
    createdAt: string
}

export type TodoList = TodoItem[];