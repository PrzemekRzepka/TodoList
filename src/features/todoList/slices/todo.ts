import { createSlice, PayloadAction } from "@reduxjs/toolkit"
import { TodoItem, TodoList } from "../../../shared/types/todoTypes"
import { RootState } from "../../../store"

const initialState: TodoList = []

const todoSlice = createSlice({
    name: "todoSlice",
    initialState,
    reducers: {
        addItem(state, action: PayloadAction<TodoItem>) {
            if (action.payload.text !== '') {
                state.push(action.payload)
            } else {
                console.log('empty item')
            }
        },

        addItems(state, action: PayloadAction<TodoList>) {
            return action.payload
        },

        removeItem(state, action: PayloadAction<string>) {
            const itemIndex = state.findIndex(item => item.id === action.payload);

            if (itemIndex === -1) {
                console.log(`Dont found item with id: ${action.payload}`)
                return state
            }
            return state.filter(item => item.id !== action.payload)
        }
    }
})


export const { addItem, addItems, removeItem } = todoSlice.actions
export const selectTodoListItems = (state: RootState) => state.todoSlice
export default todoSlice.reducer;