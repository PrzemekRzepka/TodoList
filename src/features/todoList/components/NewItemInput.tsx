import { View, TextInput, Button, StyleSheet } from "react-native"
import { useState } from "react"
import { useDispatch } from "react-redux"

import { addItem } from "../slices/todo"
import { ADD_TEXT, TEXT_INPUT_PLACEHOLDER_TEXT } from "../../../shared/types/consts"


export default function NewItemInput() {
    const [itemText, setItemText] = useState<string>('')

    const dispatch = useDispatch()

    function handleAddingItem() {
        if (itemText.trim()) {
            dispatch(addItem({ id: Date.now().toString(), text: itemText }))
        } else {
            //TODO add dialog box
            console.log('Empty note!');
        }
    }

    return (
        <View style={styles.container}>
            <TextInput
                style={styles.input}
                placeholder={TEXT_INPUT_PLACEHOLDER_TEXT}
                onChangeText={setItemText}
                value={itemText}
            />
            <View style={styles.buttonContainer}>
                <Button
                    title={ADD_TEXT}
                    onPress={handleAddingItem} />
            </View>
        </View>
    )
}

const styles = StyleSheet.create({
    container: {
        alignItems: 'flex-start',
        gap: 16

    },
    buttonContainer: {
        width: "20%"
    },
    input: {
        width: '100%',
        borderWidth: 1,
        color: "black"
    }
})