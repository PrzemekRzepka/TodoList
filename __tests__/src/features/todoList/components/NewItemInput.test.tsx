import { fireEvent, render, waitFor } from "@testing-library/react-native";
import { useDispatch } from "react-redux";

import NewItemInput from "../../../../../src/features/todoList/components/NewItemInput";
import { addItem } from "../../../../../src/features/todoList/slices/todo";
import { ADD_TEXT, TEXT_INPUT_PLACEHOLDER_TEXT } from "../../../../../src/shared/types/consts";

const TEST_TEXT = 'testText';

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
}))

describe('NewItemInput tests', () => {
    beforeEach(() => {
        ((useDispatch as unknown) as jest.Mock).mockClear();
    });

    it('snapshot', () => {
        const snapshot = render(<NewItemInput />)

        expect(snapshot).toMatchSnapshot()
    })



    it('should renders with corrected placeholder', () => {
        const { getByPlaceholderText } = render(<NewItemInput />)

        expect(getByPlaceholderText(TEXT_INPUT_PLACEHOLDER_TEXT)).toBeTruthy()
    })

    it('should change text', () => {
        const { getByPlaceholderText } = render(<NewItemInput />);
        const inputText = getByPlaceholderText(TEXT_INPUT_PLACEHOLDER_TEXT);

        fireEvent.changeText(inputText, TEST_TEXT);

        expect(inputText.props.value).toBe(TEST_TEXT);
    })

    it('should not call dispatch after button click', () => {
        const dispatchMock = jest.fn();
        ((useDispatch as unknown) as jest.Mock).mockReturnValue(dispatchMock)

        const { getByText } = render(<NewItemInput />);

        const button = getByText(ADD_TEXT);
        fireEvent.press(button);

        expect(dispatchMock).not.toHaveBeenCalled()
    })

    it.skip('should dispatch addItem after entering text and button click', async () => {
        const dispatchMock = jest.fn();
        ((useDispatch as unknown) as jest.Mock).mockReturnValue(dispatchMock);

        const { getByText, getByPlaceholderText } = render(<NewItemInput />)

        const inputText = getByPlaceholderText(TEXT_INPUT_PLACEHOLDER_TEXT);
        const button = getByText(ADD_TEXT);

        fireEvent.changeText(inputText, TEST_TEXT);

        await waitFor(() => {
            expect(inputText.props.value).toBe(TEST_TEXT);
        });

        fireEvent.press(button)

        expect(dispatchMock).toHaveBeenCalledWith(
            expect.objectContaining({
                type: addItem.type,
                payload: expect.objectContaining({
                    text: TEST_TEXT
                })
            })
        )
    })
})