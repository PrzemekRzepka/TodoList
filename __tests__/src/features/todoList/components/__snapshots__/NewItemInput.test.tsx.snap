// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewItemInput tests snapshot 1`] = `
<View
  style={
    {
      "alignItems": "flex-start",
      "gap": 16,
    }
  }
>
  <TextInput
    onChangeText={[Function]}
    placeholder="To do task text"
    style={
      {
        "borderWidth": 1,
        "color": "black",
        "width": "100%",
      }
    }
    value=""
  />
  <View
    style={
      {
        "width": "20%",
      }
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          [
            {},
          ]
        }
      >
        <Text
          style={
            [
              {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          ADD
        </Text>
      </View>
    </View>
  </View>
</View>
`;
